src/storage/files.c

```diff
@@ -339,11 +344,145 @@ static int password_callback(char* buf, int bufsiz, int verify, void* cb_tmp)
+            return NID_undef;
+        }
+        X509_ALGOR_get0(&aoid, &aparamtype, &aparam, pbe2->keyfunc);
+        pbenid = OBJ_obj2nid(aoid);
+        X509_ALGOR_get0(&aoid, 0, 0, pbe2->encryption);
+        encnid = OBJ_obj2nid(aoid);
+        PBE2PARAM_free(pbe2);
+        return encnid;
+    }
+
+    return NID_undef;
+}
+
+
+static int bag_get_pkey_algo(const PKCS12_SAFEBAG *bag, const char *pass, int passlen)
+{
+    // we only support a shrouded keybag
+    if(PKCS12_SAFEBAG_get_nid(bag) is_eq NID_pkcs8ShroudedKeyBag)
+    {
+        const X509_SIG *tp8;
+        const X509_ALGOR *tp8alg;
+
+        tp8 = PKCS12_SAFEBAG_get0_pkcs8(bag);
+        X509_SIG_get0(tp8, &tp8alg, 0);
+        return get_algo_nid(tp8alg);
+    }
+    return NID_undef;
+}
+
+
+static int bags_get_pkey_algo(const STACK_OF(PKCS12_SAFEBAG) *bags, const char *pass, int passlen, int desired_algo_nid)
+{
+    int i, algo_nid = NID_undef;
+    for(i = 0; i < sk_PKCS12_SAFEBAG_num(bags); ++i)
+    {
+        int bag_algo_nid = bag_get_pkey_algo(sk_PKCS12_SAFEBAG_value(bags, i), pass, passlen);
+        if(bag_algo_nid is_eq NID_undef)
+        {
+            continue;
+        }
+        if(bag_algo_nid not_eq desired_algo_nid)
+        {
+            return bag_algo_nid;
+        }
+        algo_nid = bag_algo_nid;
+    }
+    return algo_nid;
+}
+
+
+/* This function determines whether a given PKCS#12 file contains at least one private key that was encrypted using the
+   desired algorithm. */
+/* Iterates over all keybags in the file but only looks inside Shrouded keybags. Returns 1 if the desired algorithm
+   and no other was found, 0 otherwise. */
+static int algo_allowed(const PKCS12* p12, const char* pass, int passlen, int desired_algo_nid)
+{
+    STACK_OF(PKCS7) *auth_safes = 0;
+    STACK_OF(PKCS12_SAFEBAG) *bags = 0;
+    int i, bagnid;
+    int found_desired_algo = 0;
+    PKCS7 *p7 = 0;
+
+    auth_safes = PKCS12_unpack_authsafes(p12);
+
+    if(0 is_eq auth_safes)
+    {
+        return 0;
+    }
+
+    for(i = 0; i < sk_PKCS7_num(auth_safes); ++i)
+    {
+        p7 = sk_PKCS7_value(auth_safes, i);
+        bagnid = OBJ_obj2nid(p7->type);
+        if(bagnid == NID_pkcs7_data)
+        {
+            bags = PKCS12_unpack_p7data(p7);
+        }
+        else if(bagnid == NID_pkcs7_encrypted)
+        {
+            bags = PKCS12_unpack_p7encdata(p7, pass, passlen);
+        }
+        else
+        {
+            continue;
+        }
+        if(!bags)
+        {
+            sk_PKCS7_pop_free(auth_safes, PKCS7_free);
+            return 0;
+        }
+
+        int algo = bags_get_pkey_algo(bags, pass, passlen, desired_algo_nid);
+        sk_PKCS12_SAFEBAG_pop_free(bags, PKCS12_SAFEBAG_free);
+        bags = 0;
+
+        if(algo is_eq NID_undef)
+        {
+            continue;
+        }
+        if(algo is_eq desired_algo_nid)
+        {
+            found_desired_algo = 1;
+            continue;
```

Comment:
Why `continue` here?
You can `break` since all you wanted is find at least one such key algo.